Feature/ ODP nodes update: calculate and validate dependents nodes

minotogna opened the pull request:
part of #1627  and #1666

~~**TODO** Remove originalDataPoint from assessment section data store (feature available in pr #1640)~~

~~**Important** To be merged after #1684~~




https://user-images.githubusercontent.com/1138156/202180505-7aa2104b-19da-4667-a130-0c68020784b0.mov

minotogna committed:
extract calculateAndValidateDependentNodes

minotogna committed:
extract updateOriginalDataPointDependentNodes

minotogna committed:
add 1a/1b  updateOriginalDataPointDependentNodes

minotogna committed:
fix  updateOriginalDataPoint countryIso

minotogna committed:
fix deepscan

minotogna committed:
fix insert assessment

sorja commented on src/server/controller/cycleData/updateOriginalDataPointDependentNodes.ts:
```diff
@@ -0,0 +1,53 @@
+import { calculateAndValidateDependentNodes } from './persistNodeValue/calculateAndValidateDependentNodes'
+
+const dependencies: Array<{ sectionName: string; tableName: string; variableName: string }> = [
+  { sectionName: 'extentOfForest', tableName: 'extentOfForest', variableName: 'forestArea' },
```
tableName: TableNames.extentOfForest

minotogna replied:
sure. you will update this please

sorja committed:
Merge branch 'master-meta' into feature/odp-dependencies

sorja committed:
Use TableNames

sorja committed:
Fix tests: Include metacache

sorja committed:
Merge branch 'master-meta' of github.com:openforis/fra-platform into feature/odp-dependencies

sorja committed:
Merge branch 'master-meta' into feature/odp-dependencies

sorja committed:
Fix defaultMetaCache creation

assessment.metaCache initial creation requires cycle.uuid, so we initialize assessment.metaCache in cycle creation cycle/create.ts

sorja committed:
Merge branch 'feature/odp-dependencies' of github.com:openforis/fra-platform into feature/odp-dependencies

sorja committed:
assessmentSection/state: Move originalDataPointData under data[countryIso] and listen to odp changes on cell

sorja committed:
ODP nodes update: calculate and validate dependents nodes

sorja committed:
Merge branch 'master-meta' of github.com:openforis/fra-platform into feature/odp-dependencies

sorja committed:
deepscan

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint.ts:
```diff
@@ -1,6 +1,7 @@
 import { ActivityLogMessage, Assessment, Cycle, OriginalDataPoint } from '@meta/assessment'
 import { User } from '@meta/user'
 
+import { updateOriginalDataPointDependentNodes } from '@server/controller/cycleData/updateOriginalDataPointDependentNodes'
```
relative import here would be preferred

minotogna commented on src/server/controller/cycleData/updateOriginalDataPoint.ts:
```diff
@@ -16,11 +17,15 @@ export const updateOriginalDataPoint = async (
 ): Promise<OriginalDataPoint> => {
   const { assessment, cycle, originalDataPoint, user } = props
 
+  const updatedOriginalDataPoint = await OriginalDataPointRepository.update(
```
why is this outside transaction ?

sorja replied:
views dont update inside transaction so this needs to be done before, what do you suggest?
(; if we dont move it outside transaction, the updates are 1 state update behind, always showing the previous changed value and calculations)

minotogna commented on src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts:
```diff
@@ -43,17 +45,46 @@ export const validateNodeUpdates = async (props: Props, client: BaseProtocol): P
+          let isODPNode = false
+
+          // If we are handling tables 1a or 1b
+          if ([TableNames.extentOfForest, TableNames.forestCharacteristics].includes(tableName as TableNames)) {
+            // eslint-disable-next-line no-await-in-loop
+            const ODPYears = await OriginalDataPointRepository.getReservedYears(
+              { assessment, cycle, countryIso },
+              client
+            )
+
+            // We are on odp node if there is a year for current colName
+            isODPNode = ODPYears.map(String).includes(colName)
+          }
+
+          if (isODPNode) {
+            // Fetch the corresponding ODP value for the node update (used to update tables over websocket)
+            // eslint-disable-next-line no-await-in-loop
+            value = await getOriginalDataPointNode(
+              {
+                assessment,
+                cycle,
+                countryIso,
+                colName,
+                variableName,
+              },
+              client
+            )
+          } else {
+            // eslint-disable-next-line no-await-in-loop
+            const validation = await validateNode(
+              { assessment, cycle, countryIso, tableName, variableName, colName, row },
+              client
+            )
+            // eslint-disable-next-line no-await-in-loop
+            const node = await NodeRepository.updateValidation(
+              { assessment, cycle, tableName, variableName, countryIso, colName, validation },
+              client
+            )
+            value = node?.value
+          }
```
can u explain what's happening here?

sorja replied:
We fetch the ODP value for the node if the criteria is met:
- table 1a/1b
- year for odp

minotogna replied:
That I could see. But why ?
Can u explain why ?

sorja replied:
- we need to implement odp validation
- normal tables/views work on node table
- odp special case, works on origianl_data_point_data


what do you suggest?
change the repository underneath?
add odps to node table and handle the odps as nodes instead and get rid of specific 'odp' tables?

https://github.com/openforis/fra-platform/blob/97af1228b0cf66f5502b8b4200b6cb834f5fd963/src/server/repository/assessmentCycle/node/updateValidation.ts#L24

minotogna replied:
I think at the moment ODP validation happens only client side, isn't is so?
If yes, here we don't need to validate the ODP.

sorja replied:
so instead, in the parent where validateNodeUpdates is called we instead replace the found values? This one returns the non-odp values anyway, I must look deeper

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

minotogna committed:
fix updateOriginalDataPointDependentNodes transaction

minotogna committed:
Merge remote-tracking branch 'origin/feature/odp-dependencies' into feature/odp-dependencies

minotogna committed:
add websocket send message

minotogna committed:
update import

minotogna committed:
disabled odp update test

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies

mergify[bot] committed:
Merge branch 'master-meta' into feature/odp-dependencies
